Operational security guide (#44065)
mdn/content: files/en-us/web/security/defenses/operational_security/index.md

---
title: Operational security
slug: Web/Security/Defenses/Operational_security
page-type: guide
sidebar: security
---

In this guide we'll cover security practices that are not directly related to the code you write, but are concerned with how you develop, build, ship, and update your project.

> Many of these are defenses against [supply chain attacks](/en-US/docs/Web/Security/Attacks/Supply_chain_attacks), in which the attacker subverts the processes that you follow in developing the site, to introduce malicious or vulnerable code into it. For this reason, defenses against supply chain attacks are usually about securing your development process.

Reviewer: ```suggestion
Many of these are defenses against [supply chain attacks](/en-US/docs/Web/Security/Attacks/Supply_chain_attacks), in which the attacker subverts the processes you follow to develop your sites, to introduce malicious or vulnerable code into them. Defenses against supply chain attacks are usually about securing your development process.
```

We've split this guide into three sections:

- Securing your development environment: this includes practices to follow regarding the way you develop and update your own code.

- Managing dependencies: almost all software projects depend on packages written by a third party. This section lists practices to follow that reduce the risk involved when you choose to do this.

- Monitoring and response: this section lists practices that help you discover security problems in your project and respond to them.

## Securing your development environment

In this section we'll describe steps you can take to prevent attackers from compromising your project's code.

### Requiring strong authentication for project maintainers

> By gaining control of a maintainer's account an attacker can introduce malicious code or ship a malicious update of the product.

Reviewer: ```suggestion
By gaining control of a maintainer's account, an attacker can introduce malicious code or ship a malicious update of their product.
```

### Securing your configuration

Understand and apply secure settings for your tools, especially your source control system. Key protections are:

- Ensuring that pull requests (PRs) go through review and explicit approval from a code owner before they can be merged.
- Ensuring that PRs pass {{glossary("continuous integration")}} checks before they can be merged.
- Requiring that commits are signed.

See the OpenSSF's [Source Code Management Platform Configuration Best Practices](https://best.openssf.org/SCM-BestPractices/), which includes specific checklists for GitHub and GitLab.

## Managing third-party dependencies

Third-party dependencies include not only libraries and frameworks that your code uses, but all third-party tools involved in the development process, including editors, IDEs, source control systems, package managers, and build tools.

To mitigate problems with third-party dependencies, we'll discuss four practices:

1. Evaluating new dependencies
2. Updating existing dependencies
3. Maintaining a _Software Bill of Materials_ (SBOM)

> 4. Using Subresource Integrity for external scripts

Reviewer: This was kept in the Supply Chain Attacks article.

Author: Oh good catch, thank you: https://github.com/mdn/content/pull/44065/commits/21eb1426c7601da39109d15da3e3558ce398fb30.

### Evaluating new dependencies

> Before adding a new dependency, you should assess how much of a security risk it represents. You need to be confident that the dependency is actively maintained, that it has a record of fixing issues and a process for reporting and responding to security vulnerabilities.

Reviewer: ```suggestion
Before adding a new dependency, you should assess how much of a security risk it represents. You need to be confident that the dependency is actively maintained, has a record of fixing issues, and has a process for reporting and responding to security vulnerabilities.
```

For example, suppose your package.json includes a dependency named "example-dependency":

```json
{
  "name": "example-project",
  "version": "1.0.0",
  "dependencies": {
    "example-dependency": "^1.0.2"
  }
}
```

Suppose your project's build process automatically runs when your suppliers release new versions. The build process starts the build by calling `npm install`. This will fetch the latest version of "example-dependency", subject to the version range `"^1.0.2"`.

At version `1.0.2`, which is the point you added it to the project, "example-dependency" is a useful, benign package. Then an attacker takes over the account belonging to the developer of "example-dependency", and publishes a malicious `1.0.3` version. Your build process runs, installs the malicious package, and you are compromised.

All this has happened without any changes to your project's direct artifacts, or any opportunity for you to review the update and see if it looks suspicious.

> The solution to this is to use a lockfile when building your project. A lockfile is automatically generated whenever a project's dependencies are installed, and it lists the exact versions of the direct and indirect dependencies used in a project.

Reviewer: It would be good to add a sentence to the end of the paragraph saying something like "With the lockfile present, you can no longer install versions of the dependencies other than the ones listed in it." Just to make it absolutely clear how it works. Feels like that is missing at the moment.

Author: -> https://github.com/mdn/content/pull/44065/commits/c0c1d3d4d7647ee56f6477a6ca01de2ccea8f557

#### Review updates

When reviewing an update to a dependency, consider whether it's an update you want to accept:

- Read the changelog for the release, to understand what it claims to be offering (and whether you need to accept it at all, at this point).
- See if it introduces any additional dependencies.
- If possible, review the source code updates, and see if any of them are inexplicable or don't match up with the changelog.
- Consider waiting a little while before updating: often, supply chain attacks are quickly discovered by security researchers, and it is better for you if an update is found to be malicious before you have accepted it.

### Maintaining a Software Bill of Materials

To get a deeper insight into your dependencies, you can maintain a detailed inventory of them. This is called a _Software Bill of Materials_ (SBOM).

A lockfile is really a kind of SBOM: however, the term "SBOM" usually refers to a separate standard format for representing dependencies. These standards are usually both broader and deeper than a lockfile. That is:

- They can capture dependencies, such as web services, that are not represented in a lockfile.

> - They can capture additional information about each dependency, that is not represented in a lockfile.

Reviewer: ```suggestion
- They can capture additional information about each dependency that is not represented in a lockfile.
```

Using a standard format for representing an SBOM also means you can:

- Share your SBOM with third parties
- Integrate tools that can understand your SBOM for purposes such as regulatory compliance or vulnerability monitoring.

The two most common standards for representing a software bill of materials are:

- [CycloneDX](https://cyclonedx.org/), originally developed by [OWASP](https://owasp.org/).
- [SPDX](https://spdx.dev/), maintained by the [Linux Foundation](https://www.linuxfoundation.org/).

> Both these standards have good support, and you can use either to represent the SBOM for your project. SPDX was initially focused on helping products ensure that they are compliant with open source software licenses, but has added features to support security use cases. CycloneDX is a newer and more lightweight standard which was focused from the start on promoting supply chain security.

Reviewer: ```suggestion
Both these standards have good support, and you can use either to represent the SBOM for your project. SPDX was initially focused on helping products ensure compliance with open source software licenses, but has added features to support security use cases. CycloneDX is a newer and more lightweight standard focused from the start on promoting supply chain security.
```

- Services represent external APIs that software may call, for example through endpoint URIs.

Every component and service used in the product, either directly or indirectly, is represented by an object in the SBOM. The object includes information about the item including its name, version, author, license, description, {{glossary("hash_function", "hashes")}} (for components) and endpoint URIs (for services).

The SBOM also lists vulnerabilities that have been identified in the product's dependencies. Each item in the list contains information about this vulnerability, including a description, a set of [CWE](https://cwe.mitre.org/index.html) codes, mitigations, links to advisories, and the identifiers for the components or services that the vulnerability affects.

#### Creating an SBOM

You can generate an SBOM for a product using a separate tool such as [cdxgen](https://cdxgen.github.io/cdxgen/#/) or a command such as [`npm sbom`](https://docs.npmjs.com/cli/commands/npm-sbom/). An SBOM is usually generated as part of the build process, although it is possible to generate one at other stages of the software lifecycle.

#### Using an SBOM

> An SBOM enables you to implement several defenses against supply chain attacks, and we'll list three important ones here:

Reviewer: ```suggestion
An SBOM enables you to implement several defenses against supply chain attacks; the three most important ones are as follows:
```

I'd assume they are the most important ones. It would be weird if they weren't.

- **Vulnerability management**: one of the main uses for an SBOM is to respond to vulnerabilities that have been identified in your dependencies. You can use third-party tools such as OWASP's [Dependency-Track](https://dependencytrack.org/), which automate this by scanning sources of vulnerability reports such as the [NIST National Vulnerability Database](https://nvd.nist.gov/) or [GitHub Advisories](https://github.com/advisories).

> - **Integrity verification**: if the SBOM contains hashes for dependencies, it's possible to verify that the source of the component you're depending on has not been modified from its original released form.

Reviewer: ```suggestion
- **Integrity verification**: if the SBOM contains hashes for dependencies, it's possible to verify that the source of a dependency has not been modified from its original released form.
```

- **Supplier risk management**: by capturing information about the supplier of your dependencies, an SBOM can help you understand when you are depending on components or services from suppliers that are no longer considered reliable.

## Monitoring and response

Once a website is deployed, you can help detect and respond to attacks by logging relevant events and alerting project maintainers when potentially suspicious activity is detected.

You need to find a balance in which maintainers are alerted about real problems, but are not constantly being alerted by false alarms. The specific events which should be logged, and the subset of events that should trigger alarms, is dependent on the project and its threat model, but commonly includes:

> - Input validation failures: cases when user input isn't what your website expects. Input types include form input, URL parameters, or file uploads. Validation failures include input with unexpected values, formats, or lengths, or unexpected parameter names. User input that could not have been entered manually, such as a nonexistent {{htmlelement("select")}} option, is especially suspicious.

Reviewer: ```suggestion
- Input validation failures: cases when user input isn't what your website expects. Input types include form input, URL parameters, or file uploads. Validation failures include input with unexpected values, formats, lengths, or parameter names. User input that could not have been entered manually, such as a nonexistent {{htmlelement("select")}} option, is especially suspicious.
```

---
title: Operational security
slug: Web/Security/Defenses/Operational_security
page-type: guide
sidebar: security
---

> In this guide we'll cover security practices that are not directly related to the code you write, but are concerned with how you develop, build, ship, and update your project.

Reviewer 2: What do you think about also having a section that talks about incident reporting or disclosure management? In open source, this is often outlined in a security.md file. I think websites also sometimes have mechanisms in which users can contact the admins about security questions or vulnerabilities they discovered. It seems that this could be seen as part of a good operational practice.

Author: So I did think about this, and it is in the SWAG guidelines, but I wondered if it's more applicable to library/framework developers than web developers. It's all very well to have a security.md where your source lives, but how is that really findable from a website? Are we asking people to add some user-facing bit of the website pointing them there? Is it actually a common practice for websites to do this?

Reviewer 2: I see it more and more. https://joinmastodon.org/about#contact provides a dedicated email for security issue reporting, for example. I guess on many services you go through the normal support chat mechanisms but I think it might be better to provide a dedicated route for security reporting as a best practice?

Many of these are defenses against [supply chain attacks](/en-US/docs/Web/Security/Attacks/Supply_chain_attacks), in which the attacker subverts the processes you follow to develop your sites, to introduce malicious or vulnerable code into them. Defenses against supply chain attacks are usually about securing your development process.

> We've split this guide into three sections:

Reviewer 2: This is missing the "making backups" section

- Securing your development environment: this includes practices to follow regarding the way you develop and update your own code.

- Managing dependencies: almost all software projects depend on packages written by a third party. This section lists practices to follow that reduce the risk involved when you choose to do this.

- Monitoring and response: this section lists practices that help you discover security problems in your project and respond to them.

## Securing your development environment

> In this section we'll describe steps you can take to prevent attackers from compromising your project's code.

Reviewer 2: In this section, I would expect something about asset security. Like things that are of particular interest like secrets that developers need in order to work on prod or staging environments, databases, etc.

Author: Can you elaborate on this? What "something"?

Reviewer 2: I was mostly thinking about these kinds of things:

- Handling of secrets (their storage, their access [maybe not all devs get the same access to them?], prevent pushing them to a public repo, ...)
- Guardrails to not deploy production when you meant to deploy to staging for dev testing.
- Guardrails for database queries to not delete prod

I guess this could be summarized as some sort of policying and guardrailing for the assets that developers will need to work with.

An SBOM enables you to implement several defenses against supply chain attacks; the three most important ones are as follows:

- **Vulnerability management**: one of the main uses for an SBOM is to respond to vulnerabilities that have been identified in your dependencies. You can use third-party tools such as OWASP's [Dependency-Track](https://dependencytrack.org/), which automate this by scanning sources of vulnerability reports such as the [NIST National Vulnerability Database](https://nvd.nist.gov/) or [GitHub Advisories](https://github.com/advisories).
- **Integrity verification**: if the SBOM contains hashes for dependencies, it's possible to verify that the source of a dependency has not been modified from its original released form.
- **Supplier risk management**: by capturing information about the supplier of your dependencies, an SBOM can help you understand when you are depending on components or services from suppliers that are no longer considered reliable.

## Monitoring and response

Once a website is deployed, you can help detect and respond to attacks by logging relevant events and alerting project maintainers when potentially suspicious activity is detected.

> You need to find a balance in which maintainers are alerted about real problems, but are not constantly being alerted by false alarms. The specific events which should be logged, and the subset of events that should trigger alarms, is dependent on the project and its threat model, but commonly includes:

Reviewer 2: ```suggestion
You need to find a balance in which maintainers are alerted about real problems, but are not constantly being alerted by false alarms. The specific events which should be logged, and the subset of events that should trigger alarms, is dependent on the project and its [threat model](/en-US/docs/Web/Security/Threat_modeling), but commonly includes:
```

- Input validation failures: cases when user input isn't what your website expects. Input types include form input, URL parameters, or file uploads. Validation failures include input with unexpected values, formats, lengths, or parameter names. User input that could not have been entered manually, such as a nonexistent {{htmlelement("select")}} option, is especially suspicious.

- Events related to authentication:
  - Failed attempts to sign in, especially repeated failures in a short space of time.
  - Logins from unexpected locations or devices.
  - Credential management events: changing, creating, or deleting credentials.
  - Triggering secondary flows such as forgotten password flows.

- Access control failures: attempts to access resources without the correct level of authorization.

> - [CSP](/en-US/docs/Web/HTTP/Guides/CSP) violations.

Reviewer 2: I think the Reporting APIs deserve some mentioning in this section or at least a link. https://developer.mozilla.org/en-US/docs/Web/API/Reporting_API

## Making backups

Making regular backups of your site and, especially, your data, provides a defense against a number of problems, including but not limited to active attacks:

- Errors by maintainers or buggy tools resulting in data loss or corruption.
- Vandalism by an attacker with write access.
- [Ransomware](https://en.wikipedia.org/wiki/Ransomware) attacks, in which an attacker makes the victim's data inaccessible (for example, by encrypting it) unless the victim pays a ransom to recover it.

> 

Reviewer 2: Should we say something about the security of the backup? If attackers gain access to backups, it can be quite terrible, right?

Author: Yeah, I suppose so.

I don't especially like this section already, because it feels handwavy - it tells people to do something without really telling them how to do it. But a proper guide to making and maintaining backups is a whole guide in itself. I suppose I didn't talk about backup security because it seems obvious, and saying "btw your backups should also be secure" doesn't seem very helpful without explaining what that means and how to do it, and we don't really propose to do this here.

- Input validation failures: cases when user input isn't what your website expects. Input types include form input, URL parameters, or file uploads. Validation failures include input with unexpected values, formats, lengths, or parameter names. User input that could not have been entered manually, such as a nonexistent {{htmlelement("select")}} option, is especially suspicious.

- Events related to authentication:
  - Failed attempts to sign in, especially repeated failures in a short space of time.
  - Logins from unexpected locations or devices.
  - Credential management events: changing, creating, or deleting credentials.
  - Triggering secondary flows such as forgotten password flows.

- Access control failures: attempts to access resources without the correct level of authorization.

- [CSP](/en-US/docs/Web/HTTP/Guides/CSP) violations, which can be reported using the [Reporting API](/en-US/docs/Web/API/Reporting_API).

> You should also proviude a way for users to report security problems with your site. Projects can communicate their security policy using a `SECURITY.md` file in the root directory of their repository. This file explains how users or researchers can report vulnerabilities, goe you will handle these reports, and details you your bug bounty program, if you have one.

Reviewer 2: ```suggestion
You should also provide a way for users to report security problems with your site. Projects can communicate their security policy using a `SECURITY.md` file in the root directory of their repository. This file explains how users or researchers can report vulnerabilities, how you will handle these reports, and details of your bug bounty program, if you have one.
```

Author: OK I'm not sure what happened to me there.
